cpp/src/arrow/array/array_base.h

```diff
@@ -232,15 +233,32 @@ class ARROW_EXPORT Array {
   /// \return DeviceAllocationType
   DeviceAllocationType device_type() const { return data_->device_type(); }
 
+  /// \brief Return the statistics of this Array
+  ///
+  /// \return const std::shared_ptr<ArrayStatistics>&
+  const std::shared_ptr<ArrayStatistics>& statistics() const { return statistics_; }
+
  protected:
   Array() = default;
   ARROW_DEFAULT_MOVE_AND_ASSIGN(Array);
 
   std::shared_ptr<ArrayData> data_;
   const uint8_t* null_bitmap_data_ = NULLPTR;
 
-  /// Protected method for constructors
-  void SetData(const std::shared_ptr<ArrayData>& data) {
+  /// Protected method for constructors. This must be called from each
+  /// array class to call its SetData(). You can't use call this in a
+  /// parent array class.
+  void Init(const std::shared_ptr<ArrayData>& data,
+            const std::shared_ptr<ArrayStatistics>& statistics) {
+    SetData(data);
+    if (statistics) {
+      SetStatistics(statistics);
+    }
+  }
+
+  /// Protected method for constructors. Don't call this method
+  /// directly. This should be called from Init().
+  virtual void SetData(const std::shared_ptr<ArrayData>& data) {
```

Comment:
> This probably shouldn't be virtual since it's called from constructors, [which adds nuances.](https://en.cppreference.com/w/cpp/language/virtual#During_construction_and_destruction) Specifically, if `BaseArray::BaseArray()` calls `Array::Init()` then that will not call `DerivedArray::SetData()`.

You're right. So this implementation ensures calling `Array::Init()` from `DerivedArray::DerivedArray()` not `BaseArray::BaseArray()`. It calls `DerivedArray::SetData()` but this is a tricky limitation. I'll consider another approach.